walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,69 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
+
 
 public final class FindMeetingQuery {
+
+  private interface Criterion {
+    boolean satisfies(int x);
+  }
+
+  private static final int DAY = 60 * 24;
+
+  private int getStart(Event event) {
+    return event.getWhen().start();
+  }
+
+  private int getEnd(Event event) {
+    return event.getWhen().end();
+  }
+
+  private boolean sameAttendees(Collection<String> a, Collection<String> b) {
+    return new HashSet<>(a).removeAll(b);
+  }
+
+  private List<TimeRange> makeSchedule(int [] occupied, Criterion criterion, int duration) {
+    List<TimeRange> schedule = new ArrayList<>();
+    int index = 0;
+    while (index < DAY) {
+      int start = index;
+      while (index < DAY && criterion.satisfies(occupied[index])) {
+        index++;
+      }
+      if (index - start >= duration) {
+        schedule.add(new TimeRange(start, index - start));
+      }
+      while (index < DAY && !criterion.satisfies(occupied[index])) {
+        index++;
+      }
+    }
+    return schedule;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
```

Comment:
for public methods, do some sanity checks that the passed arguments make sense first, e.g. check for when `events` is `null` or when `request` is null and throw an appropriate `IllegalArgumentException` there. Look at how TimeRange.java is implemented for example, as a reference.
Note that for private methods it's OK to not always check all the bits and pieces, because you control all call sites (however it's a good practice and may help find bugs earlier).